### PR TITLE
Fixes clang conversion warnings

### PR DESCRIPTION
**Pull request by TinyTinni**
Hi,
when I compile duckdb with clang 16/17.06, I get several conversion warnings like the following:

> duckdb/src/include/duckdb/common/types/datetime.hpp:27:10: warning: implicit conversion from 'const int64_t' (aka 'const long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
[build]    27 |                 return micros;
[build]       |                 ~~~~~~ ^~~~~~
[build] 1 warning generated.

Since this warning is in a header file, it occurs sadly a lot during compilation. The log gets a bit confusing with all these false positives.

This PR silences some of the warnings by making the cast explicit. It is focused on headers and templates which generate most of it (especially the one in the example "datetime.hpp"). It reduces the warning count from 529 to 129.

The remaining warnings seems to be better suited for another PR, as they are kinda unrelated, except sharing the same warning category.

I also tried to achieve the same numerical stability as before.

### CODE REVIEW

File: src/core_functions/aggregate/distributive/entropy.cpp [hunk 1]
```diff
@@ -50,11 +50,12 @@ struct EntropyFunctionBase {
 
 	template <class T, class STATE>
 	static void Finalize(STATE &state, T &target, AggregateFinalizeData &finalize_data) {
-		double count = state.count;
+		double count = static_cast<double>(state.count);
```
Comment (hawkfish) [inline src/core_functions/aggregate/distributive/entropy.cpp:53]: We have wrappers for these static casts called `NumericCast` in `numeric_utils.hpp`. They will trap conversion errors, so can we switch to using them?

File: src/include/duckdb/common/types/datetime.hpp [hunk 2]
```diff
@@ -24,7 +24,7 @@ struct dtime_t { // NOLINT
 		return micros;
 	}
 	explicit inline operator double() const {
-		return micros;
+		return static_cast<double>(micros);
```
Comment (hawkfish) [inline src/include/duckdb/common/types/datetime.hpp:27]: `NumericCast`

File: src/include/duckdb/core_functions/aggregate/algebraic/covar.hpp [hunk 3]
```diff
@@ -53,15 +53,17 @@ struct CovarOperation {
...
+			const auto meanx = (source_count * source.meanx + target_count * target.meanx) / count;
+			const auto meany = (source_count * source.meany + target_count * target.meany) / count;
 
 			//  Schubert and Gertz SSDBM 2018, equation 21
 			const auto deltax = target.meanx - source.meanx;
 			const auto deltay = target.meany - source.meany;
-			target.co_moment =
-			    source.co_moment + target.co_moment + deltax * deltay * source.count * target.count / count;
+			target.co_moment = source.co_moment + target.co_moment +
+			                   deltax * deltay * static_cast<double>(source.count * target.count) / count;
```
Comment (hawkfish) [inline src/include/duckdb/core_functions/aggregate/algebraic/covar.hpp:66]: We have our own formatter (`make format-head`) - did you editor do this?

File: src/include/duckdb/core_functions/aggregate/algebraic/stddev.hpp [hunk 4]
```diff
@@ -49,7 +49,8 @@ struct STDDevBaseOperation {
 	}
 
 	template <class INPUT_TYPE, class STATE, class OP>
-	static void ConstantOperation(STATE &state, const INPUT_TYPE &input, AggregateUnaryInput &unary_input, idx_t count) {
+	static void ConstantOperation(STATE &state, const INPUT_TYPE &input, AggregateUnaryInput &unary_input,
```
Comment (hawkfish) [inline src/include/duckdb/core_functions/aggregate/algebraic/stddev.hpp:52]: Looks like your editor has changed this too?

**Review by hawkfish — COMMENTED** [review]
Thanks for the PR! Have a look at our cast wrappers and code formatters, but otherwise it looks reasonable to m.

**Comment by TinyTinni** [comment]
Hi,

thanks for your feedback. I will fix the formatting. Yes, my editor did this according to the checked in .clang-format. Maybe I use another clang-format version.

Regarding NumericCast.

I see a problem here. The casts are from an integer to a double.
When I look at NumericCast, it has some logic based on signed/unsigned [1].

Furthermore, the code tries to create an unsigned object it wants to cast to, which is a double. That is not defined [2] and results in a compiler error.

I could add another path, something like right after the "if (is_same<TO, FROM>)...":
```cpp
if (std::is_floating_point<TO>::value && std::is_numeric<FROM>::value)
{
    return static_cast<TO>(val);
} 
```
Thinking about it, I am unsure if this is the right solution. While this solution would mirror the status quo without warning but perhaps could improved.


[1] https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/numeric_utils.hpp#L77
[2] https://en.cppreference.com/w/cpp/types/make_unsigned

**Review by hawkfish — APPROVED** [review]
After consulting with @Mytherin this is fine. We will probably add a `LossyCast` template at some point so this can be more explicit.

**Comment by TinyTinni** [comment]
Hi,

Thanks. I hopefully fixed the formatting. It is indeed a version problem, you use clang-format 11 and I have locally clang-format 17. Couldn't install clang-format 11 here, so I did a manual edit.

Thinking about your future plans, which sound nice, it might be a good idea to rethink this PR.
This PR removes the warnings which could be very convenient when introducing the `LossyCast`.

I would totally understand if you reject this PR. I can silence the warnings locally. This would be totally fine for me. Of course, merging would be also fine.

**Comment by Mytherin** [comment]
Thanks! We'll merge it for now.

**Comment by carlopi** [comment]
Hi! I noticed this PR introduced some changes that turned out breaking the OSX CI (for example https://github.com/duckdb/duckdb/actions/runs/9491670020/job/26157537491#step:6:204), I have a PR basically ready with a fix, if you want to take a second look, I will tag you.

Problem is not only with the check, but the fact that the changes in the covariance code here: https://github.com/duckdb/duckdb/pull/12467/files#diff-c68cada9f69d55d407629f615eca24db6800d11daf9cf1fc467a1dd3593fcf26 changed semantic, since rountrip uint64 -> double -> uint64 is not guaranteed to keep precision.

The same problem is also in the stddev code, there I have not written a fix yet, your call if you want to do it.

**Comment by carlopi** [comment]
PR is here: https://github.com/duckdb/duckdb/pull/12515

File: src/include/duckdb/core_functions/aggregate/algebraic/covar.hpp (hunk 3 above)
Comment (carlopi) [reply src/include/duckdb/core_functions/aggregate/algebraic/covar.hpp]: Found the problem here, our formatter for some reason is not run on that file, I will add them back.

**Pull request merged**